src/lib/authHandler.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import * as core from '@actions/core';
+import { Runner } from '.';
+
+export class AuthHandler {
+
+
+    private _envUrl!: string;
+    private _username!: string;
+    private _password!: string;
+    private _appId!: string;
+    private _clientSecret!: string;
+    private _tenantId!: string;
+    private _pac: Runner;
+
+    constructor(private readonly pac: Runner) {
+        this._pac = pac;
+    }
+
+    public async authenticate(authKind : AuthKind) : Promise<void> {
+        core.startGroup('authentication')
+        this._envUrl = core.getInput('environment-url', { required: false });
+        const authType = this.determineAuthType();
+
+        if (authType === AuthTypes.USERNAME_PASSWORD) {
+            await this.authenticateWithUsernamePassword(authKind);
+        } else if (authType == AuthTypes.APPID_SECRET) {
+            await this.authenticateWithClientCredentials(authKind);
+        } else {
+            throw new Error('Must provide either username/password or client credential for authentication!');
+        }
+        core.endGroup();
+    }
+
+    private determineAuthType(): AuthTypes {
+
+        const validUsernameAuth = this.isValidUsernameAuth();
+        const validSPNAuth = this.isValidSPNAuth();
+
+        try{
+            if (validUsernameAuth && validSPNAuth) {
+                throw new Error('Must pick either username/password or client credential as the authentication flow.');
+            }
+
+            if (validUsernameAuth) {
+                return AuthTypes.USERNAME_PASSWORD;
+            }
+            else if (validSPNAuth) {
+                return AuthTypes.APPID_SECRET;
+            }
+        } catch (error){
+            core.setFailed(`failed: ${error.message}`);
+            throw error;
+        }
+
+        return AuthTypes.INVALID_AUTH_TYPE;
+    }
+
+    private isValidUsernameAuth(): boolean {
+        this._username = core.getInput('user-name', { required: false });
+        if (this._username) {
+            this._password = core.getInput('password-secret', { required: true});
```

Comment:
won't this throw if the password-secret is indeed missing? The way how this method is used is to probe for what info has been supplied and then turn that into a proper and friendlier error message

Reply:
Right, it would throw if password is missing. I just thought I would do an early conditional check here itself that if username is provided, password is required. Similarly below for SPN, if appId is provided, clientSecret and tenantId is required. I can remove these additional checks as well and let the user see the broader error message as "_Must provide either username/password or client credential for authentication!_"